src/LibGPS.lua

```diff
@@ -93,15 +93,71 @@ local function GetPlayerWaypoint()
     return LMP:GetMapPing(MAP_PIN_TYPE_PLAYER_WAYPOINT)
 end
 
-local function SetMeasurementWaypoint(x, y)
+local function ClearCurrentWaypoint()
+    currentWaypointX, currentWaypointY = 0, 0, nil
+end
+
+local function RestoreCurrentWaypoint()
+   if(not currentWaypointMapId) then
+        LogMessage(LOG_DEBUG, "Called RestoreCurrentWaypoint without stored waypoint.")
+        return
+    end
+
+    local wasSet = false
+    if (currentWaypointX ~= 0 or currentWaypointY ~= 0) then
+        wasSet = SetPlayerWaypointByWorldLocation(currentWaypointX, 1, currentWaypointY)
+        if (not wasSet) then
+            LogMessage(LOG_DEBUG, "Cannot reset waypoint because it was outside of our reach")
+        end
+    end
+
+    if(wasSet) then
+        LogMessage(LOG_DEBUG, "Waypoint was restored, request pin update")
+        needWaypointRestore = true -- we need to update the pin on the worldmap afterwards
+    else
+        RemovePlayerWaypoint()
+    end
+
+    ClearCurrentWaypoint()
+end
+
+local function SetMeasurementWaypoint(localX, localY)
+    local hasWaypoint = LMP:HasMapPing(MAP_PIN_TYPE_PLAYER_WAYPOINT)
+    if hasWaypoint then
+        currentWaypointX, currentWaypointY = GetPlayerWaypoint()
+        currentWaypointMapId = GetMapTileTexture()
+    end
+
     -- this waypoint stays invisible for others
     lib.suppressCount = lib.suppressCount + 1
     LMP:SuppressPing(MAP_PIN_TYPE_PLAYER_WAYPOINT)
-    LMP:SetMapPing(MAP_PIN_TYPE_PLAYER_WAYPOINT, MAP_TYPE_LOCATION_CENTERED, x, y)
-end
 
-local function SetPlayerWaypoint(x, y)
-    LMP:SetMapPing(MAP_PIN_TYPE_PLAYER_WAYPOINT, MAP_TYPE_LOCATION_CENTERED, x, y)
+    local _, pwx, pwh, pwy = GetUnitWorldPosition("player")
+    local x, y = 2000000, 2000000
+    if not SetPlayerWaypointByWorldLocation(x, pwh, y) then
+        x = -x
+        if not SetPlayerWaypointByWorldLocation(x, pwh, y) then
+            y = -y
+            if not SetPlayerWaypointByWorldLocation(x, pwh, y) then
+                x = -x
+                if not SetPlayerWaypointByWorldLocation(x, pwh, y) then
+                    LogMessage(LOG_WARNING, "Cannot set reference waypoint")
+                    ClearCurrentWaypoint()
+                    return 0, 0, false
+                end
+            end
+        end
+    end
+    local wpX, wpY = GetPlayerWaypoint()
+
+    local dwx, dwy = x - pwx, y - pwy
+    local dnx, dny = wpX - localX, wpY - localY
+    local scale = math.sqrt((dwx*dwx+dwy*dwy)/(dnx*dnx+dny*dny))
+
+    if hasWaypoint then
+        currentWaypointX, currentWaypointY = (currentWaypointX-localX) * scale + pwx, (currentWaypointY-localY) * scale + pwy

```

Comment:
Not sure how I feel about this. Above we assign the result from GetPlayerWaypoint to currentWaypoint and here we transform it, meaning we have a completely unrelated value stored in the same variable for a time. Maybe keep the normalized waypoint coordinates in a separate variable and rename this to currentWaypointWorld?

Reply:
But within the same function. Better as it was before.

Reply:
If v2 this is used by the local function RestoreWaypoint, which nobody else can use. The SaveCurrentWaypoint has been removed. So, SetReferenceWaypoint is the SaveCurrentWaypoint.
In v3 the save/restore is more public.
If SaveCurrentWaypoint also saves the current player position and the zone WorldUnit scale factor is saved beyond SetReferenceWaypoint (which we need anyway in v3), the RestoreWaypoint can work independ.

Reply:
I was referring to the hypothetical situation where someone is changing code and doesn't realize that the variable can contain a value of a different type (normalized vs world coords), which could end up quite confusing. But I suppose I'm just overthinking things.